### PR TITLE
Initial commit of the PCS storage layer.

### PR DESCRIPTION
**Pull request by mpkelly-hpe**
This layer contains an interface for PCS storage operations, storage
being an external data store that is independent of all PCS instances.
This layer currently includes an ETCD and an in-memory implementation,
both based on hms-hmetcd.

In the future we may also add support for memcached.

This was tested locally, via new unit tests, and via the PCS integration
tests.   Note that hms-hmetcd has been tested to death, so the only
real testing was to insure our calls to hms-hmetcd work correctly.

### CODE REVIEW

File: internal/storage/distlock_etcd_impl.go [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+		d.Logger = logrus.New()
+	}
+}
+
+func (d *ETCDLockProvider) Ping() error {
+	e := toStorageETCD(d)
+	return e.Ping()
+}
+
+func (d *ETCDLockProvider) DistributedTimedLock(maxSeconds int) error {
```
Comment (nieuwsma) [inline internal/storage/distlock_etcd_impl.go:76]: I could use a walk though on the distributed lock stuff.

**Review by nieuwsma — CHANGES_REQUESTED** [review]

File: docker-compose.developer.environment.yaml [hunk 2]
```diff
@@ -1,4 +1,4 @@
-version: '3.7'
+version: '3.3'
```
[the remark this answers is not in the document]
Comment (mpkelly-hpe) [reply docker-compose.developer.environment.yaml:1]: Crap.  An artifact of my testing.  I'll change it back.

File: internal/storage/distlock_mem_impl.go [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+//This file contains an in-memory implementation of a distributed locking
+//mechanism.   NOTE: THIS MECHANISM DOESN'T REALLY DO ANY DIST'D LOCKING,
+//since there is no IPC or actual backing store.  It is a wrapper around the
+//ETCD locking mechanism, which contains an in-memory implementation.  This
+//implementation is just to satisfy the dist'd lock interface.
+
+
+type MEMLockProvider struct {
+	Logger *logrus.Logger
+	Duration int
```
[the remark this answers is not in the document]
Comment (mpkelly-hpe) [reply internal/storage/distlock_mem_impl.go:41]: Just trying to KISS.   I can't imagine any need to have a timed lock duration that needs less than
second granularity.

[the remark this answers is not in the document]
Comment (nieuwsma) [reply internal/storage/distlock_mem_impl.go:41]: IMO this is a bit of an anti-pattern.  I think we should use the https://pkg.go.dev/time#Duration built in.  That is the idiomatic GO way of doing it.

[the remark this answers is not in the document]
Comment (nieuwsma) [reply internal/storage/distlock_mem_impl.go:41]: I agree that its very unlikely that we would ever want something less than 1 second; but by using Go's 'duration' as part of Time we always have a standard unit and any 'math' we need off of it is very simple.   they have built in multipliers like time.second.

File: docker-compose.developer.environment.yaml (hunk 2 above)
[the remark this answers is not in the document]
Comment (mpkelly-hpe) [reply docker-compose.developer.environment.yaml]: Fixed.

File: internal/storage/distlock_mem_impl.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mpkelly-hpe) [reply internal/storage/distlock_mem_impl.go]: Changed to use time.Duration.

**Review by nieuwsma — APPROVED** [review]

File: go.mod [hunk 4]
```diff
@@ -1,12 +1,15 @@
 module github.com/Cray-HPE/hms-power-control
 
 go 1.16
-	//todo hms-base needs to be converted to hms-xname as soon as that package is available.
+
+//todo hms-base needs to be converted to hms-xname as soon as that package is available.
 
 require (
 	github.com/Cray-HPE/hms-base v1.15.1
```
Comment (rsjostrand-hpe) [inline go.mod:8]: We should updated the hms-base package to v2.0.0, so we get all of our xname related types from hms-xname.

Comment (nieuwsma) [reply go.mod:8]: agree; I will do this in our integration.

**Pull request merged**